### PR TITLE
The Witness: Change Regions, Areas and Connections from Dict[str, Any] to dataclasses&NamedTuples

### PR DESCRIPTION
**Pull request by NewSoupVi**
This should make the code more readable, adds more typing, and makes steps towards removing the .txt Logic files (although they will still always be kept as a "source of truth")

I also wanna do Entities, but I thought this would be plenty to review for one PR

Also rename some stuff to be more consistent, mainly change "lambda" to "rule" or "requirement"

Fixes an optional requirement of https://github.com/ArchipelagoMW/Archipelago/pull/4410

Tested:
Running ruff and mypy for correct typing and linting and stuff
Unit tests should hopefully catch any regressions with generation? They are quite extensive nowadays and cover a lot of weird edge cases

### CODE REVIEW

**Review by Jouramie — APPROVED** [review]
Code LGTM, just some nitpicking.

File: worlds/witness/data/definition_classes.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from dataclasses import dataclass, field
+from typing import FrozenSet, List, NamedTuple
+
+# A WitnessRule is just an or-chain of and-conditions.
+# It represents the set of all options that could fulfill this requirement.
+# E.g. if something requires "Dots or (Shapers and Stars)", it'd be represented as: {{"Dots"}, {"Shapers, "Stars"}}
+# {} is an unusable requirement.
+# {{}} is an always usable requirement.
+WitnessRule = FrozenSet[FrozenSet[str]]
```
[the remark this answers is not in the document]
Comment (NewSoupVi) [reply worlds/witness/data/definition_classes.py:9]: Fair

File: worlds/witness/data/definition_classes.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+
+@dataclass
+class RegionDefinition:
+    name: str
+    short_name: str
+    logical_entities: List[str] = field(default_factory=list)
+    physical_entities: List[str] = field(default_factory=list)
```
[the remark this answers is not in the document]
Comment (NewSoupVi) [reply worlds/witness/data/definition_classes.py:17]: I don't want it to be inconsistent in the world and I can't change all of them now because it would create millions of conflicts

**Review by Exempt-Medic — APPROVED** [review]

**Comment by NewSoupVi** [comment]
Fuck me
